src/EntityComponentManager.cc

```diff
@@ -1106,49 +1108,41 @@ void EntityComponentManager::SetState(
         continue;
       }
 
-      // Create component
-      auto newComp = components::Factory::Instance()->New(compMsg.type());
-
-      if (nullptr == newComp)
-      {
-        ignerr << "Failed to deserialize component of type [" << compMsg.type()
-               << "]" << std::endl;
-        continue;
-      }
-
-      std::istringstream istr(compMsg.component());
-      newComp->Deserialize(istr);
-
-      // Get type id
-      auto typeId = newComp->TypeId();
-
-      // TODO(louise) Move into if, see TODO below
-      this->RemoveComponent(entity, typeId);
-
       // Remove component
       if (compMsg.remove())
       {
+        this->RemoveComponent(entity, type);
         continue;
       }
 
       // Get Component
-      auto comp = this->ComponentImplementation(entity, typeId);
+      auto comp = this->ComponentImplementation(entity, type);
 
       // Create if new
       if (nullptr == comp)
       {
-        this->CreateComponentImplementation(entity, typeId, newComp.get());
+        auto newComp = components::Factory::Instance()->New(type);
+        if (nullptr == newComp)
+        {
+          ignerr << "Failed to create component type ["
+            << compMsg.type() << "]" << std::endl;
+          continue;
+        }
+        std::istringstream istr(compMsg.component());
+        newComp->Deserialize(istr);
+        this->CreateComponentImplementation(entity, type, newComp.get());
       }
       // Update component value
       else
       {
-        ignerr << "Internal error" << std::endl;
-        // TODO(louise) We're shortcutting above and always  removing the
-        // component so that we don't get here, gotta figure out why this
-        // doesn't update the component. The following line prints the correct
-        // values.
-        // igndbg << *comp << "  " << *newComp.get() << std::endl;
-        // *comp = *newComp.get();
+        std::istringstream istr(compMsg.component());
+        comp->Deserialize(istr);
+        // TODO(anyone) mark the component as a one time or periodic change?
+        // This is done for the SetState method with a msgs::SerializedStateMap
+        // type (see method below), but nothing is currently being done for a
+        // msgs::SerializedState type. In ign-gazebo6, there's an
+        // AddModifiedComponent method that takes care of this, but that method
+        // was introduced in ign-gazebo4 (see ign-gazebo PR #742)
```

Comment:
Can we handle it the same way that's done for the `*Map` message? I think the reason that was only added to the map is that we haven't been using the non-map version in the `SceneBroadcaster`.

Reply:
Yeah, I went ahead and updated this in https://github.com/ignitionrobotics/ign-gazebo/pull/1149/commits/3ab37effcfaf16f4e556fa0f2244c5c4b99b47a6. I also rebased since this is a backport. If everything else looks good to you, feel free to merge once CI is green!